builtin/common/datastructures.lua

```diff
@@ -0,0 +1,317 @@
+local stack_mt
+stack_mt = {
+	__index = {
+		push = function(self, v)
+			self.n = self.n+1
+			self[self.n] = v
+		end,
+		pop = function(self)
+			local v = self[self.n]
+			self[self.n] = nil
+			self.n = self.n-1
+			return v
+		end,
+		is_empty = function(self)
+			return self.n == 0
+		end,
+		size = function(self)
+			return self.n
+		end,
+		clone = function(self, copy_element)
+			local stack
+			if copy_element then
+				stack = {n = self.n, true}
+				for i = 1, self.n do
+					stack[i] = copy_element(self[i])
+				end
+			else
+				stack, n = self:to_table()
+				stack.n = n
+			end
+			setmetatable(stack, stack_mt)
+			return stack
+		end,
+		to_table = function(self)
+			local t = {}
+			for i = 1, self.n do
+				t[i] = self[i]
+			end
+			return t, self.n
+		end,
+		to_string = function(self, value_tostring)
+			if self.n == 0 then
+				return "empty stack"
+			end
+			value_tostring = value_tostring or tostring
+			local t = {}
+			for i = 1, self.n do
+				t[i] = value_tostring(self[i])
+			end
+			return self.n .. " elements; bottom to top: " ..
+				table.concat(t, ", ")
+		end,
+	}
+}
+
+function core.create_stack(data)

```

Comment:
Why not `input, n` and then set `n = n or input.n`? Wouldn't that be more efficient?

Reply:
core.create_stack does not need to be as efficient as possible, it's only invoked when creating a stack, an extendable parameter list passed as table is more desirable.